slatedb/src/transactional_object/mod.rs

```diff
@@ -0,0 +1,919 @@
+    fn object(&self) -> &T;
+
+    /// Returns a `DirtyObject` with the current version ID and object which can be
+    /// modified locally and passed to `update` to persist mutations durably.
+    fn prepare_dirty(&self) -> Result<DirtyObject<T, Id>, TransactionalObjectError>;
+
+    /// Refreshes the in-memory view of the object with the latest version stored durably.
+    /// This may result in a different in-memory view returned by `object` (and different id
+    /// returned by `id`) if another process has successfully updated the object.
+    async fn refresh(&mut self) -> Result<&T, TransactionalObjectError>;
+
+    /// Transactionally update the object. Will succeed iff the version id in durable storage
+    /// matches the version id of the provided `DirtyObject`. If the versions don't match
+    /// then this fn returns `ObjectVersionExists`.
+    async fn update(&mut self, dirty: DirtyObject<T, Id>) -> Result<(), TransactionalObjectError>;
+
+    /// Transactionally update the object using the supplied mutator, if the mutator returns
+    /// `Some`. This fn will indefinitely retry the mutation on a version conflict by refreshing
+    /// and re-applying the mutation.
+    async fn maybe_apply_update<F, Err>(
+        &mut self,
+        mutator: F,
+    ) -> Result<(), TransactionalObjectError>
+    where
+        Err: std::error::Error + Send + Sync + 'static,
+        F: Fn(&Self) -> Result<Option<DirtyObject<T, Id>>, Err> + Send + Sync,
+    {
+        loop {
+            let Some(dirty) = mutator(self).map_err(|e| CallbackError(Box::new(e)))? else {
+                return Ok(());
+            };
+            match self.update(dirty).await {
+                Err(TransactionalObjectError::ObjectVersionExists) => {
+                    self.refresh().await?;
+                    continue;
+                }
+                Err(e) => return Err(e),
+                Ok(()) => return Ok(()),
+            }
+        }
+    }
+}
+
+/// Wraps `SimpleTransactionalObject` with epoch-based fencing to provide mutually-exclusive
+/// access to the object. When creating a `FenceableTransactionalObject` the caller supplied
+/// `get_epoch` and `set_epoch` fns for getting and setting the epoch in the contained object.
+/// The epoch is a monotonically increasing u64. `set_epoch` is called from
+/// `FenceableTransactionalObject#init` to set the epoch to the next value. Once the epoch is set
+/// it is never reset. Before any update, and after every refresh, this type checks whether the
+/// epoch stored in the object is higher than the epoch stored in `init`. If it is, then the
+/// corresponding `update` or `refresh` fails with`Fenced`.
+pub(crate) struct FenceableTransactionalObject<T: Clone, Id: Copy = MonotonicId> {
+    delegate: SimpleTransactionalObject<T, Id>,
+    local_epoch: u64,
+    get_epoch: fn(&T) -> u64,
+}
+
+impl<T: Clone + Send + Sync> FenceableTransactionalObject<T, MonotonicId> {
+    pub(crate) async fn init(
```

Comment:
I don't totally get this API - this seems to me to be part of the protocol more than part of the object. I might get it more when I look through example usages, but it seems like the fencing should always happen whenever I attemt to write a new version, not just when I init it?

Reply:
yeah I noticed that after leaving this comment 😅 meant to resolve this one in favor of the one I left below